<a id="0"></a>

<div <h1 <strong style="color:#F7C04A;font-size:55px;font-family:Georgia;text-align:center;">RATING PRODUCTS </strong></div>
<div <h1      <strong style="color:#F7C04A;font-size:45px;font-family:Georgia;text-align:center;">SORTING PRODUCTS</strong></div>
<div <h1    <strong style="color:#F7C04A;font-size:55px;font-family:Georgia;text-align:center;">SORTING REVIEWS </strong></div>
<div <h1    <strong style="color:#F7C04A;font-size:45px;font-family:Georgia;text-align:center;">⭐⭐⭐⭐⭐ </strong> </div>
             
<div style="float:center;margin-left:3px;max-width:100%;">
<img src="https://i.ibb.co/VMdGsbj/image-processing20210912-30371-a7o0gh.gif"></div>

<div <h1    <strong style="color:#F7C04A;font-size:60px;font-family:Georgia;text-align:center;">⭐⭐⭐⭐⭐ </strong> </div>


<div <h1 <strong style="color:#3795BD;font-size:30px;font-family:Georgia;text-align:center;">“When a measure becomes a target, it stops being a good measure.”</strong></div>
         
     
      

<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p></p><div class="list-group" id="list-tab" role="tablist">
  <h3 class="list-group-item list-group-item-action active" data-toggle="list" role="tab" aria-controls="home" style = "background-color:#000000;font-family:Georgia;color:#FFEA20;font-size:200%;text-align:LEFT;border-radius:20px 40px;overflow:hidden;border-style:dotted;border-width:1.8px;border-color:#FFEA20;">,   Table of Contents</h3>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#1" role="tab" aria-controls="home" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">User-Based Weighted Average<span class="badge badge-primary badge-pill">1</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#2" role="tab" aria-controls="messages" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Weighted Rating<span class="badge badge-primary badge-pill">2</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#3" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Bayesian Average Rating Score<span class="badge badge-primary badge-pill">3</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#4" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Sorting by Rating<span class="badge badge-primary badge-pill">4</span></a> 
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#5" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Sorting by Comment Count or Purchase Count<span class="badge badge-primary badge-pill">5</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#6" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Sorting by Bayesian Average Rating Score (Sorting Products with 5 Star Rated)<span class="badge badge-primary badge-pill">6</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#7" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">Hybrid Sorting: BAR Score + Others<span class="badge badge-primary badge-pill">7</span></a>
  <a class="list-group-item list-group-item-action" data-toggle="list" href="#9" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">^^^^^<span class="badge badge-primary badge-pill">8</span></a>
<a class="list-group-item list-group-item-action" data-toggle="list" href="#9" role="tab" aria-controls="settings" target="_self" style = "color:#000000; font-family:Charter;font-size:120%;">^^^^^<span class="badge badge-primary badge-pill">0</span></a>
</div>
</div>
</div>






<div <h1 <strong style="color:#F7C04A;font-size:20px;font-family:Georgia;text-align:center;">  Packages-Libraries-Settings 📜 ⚙️ </strong></div>


In [33]:
#Libraries
import pandas as pd
import datetime as dt
import math
import scipy.stats as st
from sklearn.preprocessing import MinMaxScaler
#Settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', 500)
pd.set_option('display.expand_frame_repr', False)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

<div <h1 <strong style="color:#3795BD;font-size:30px;font-family:Georgia;">1 - Amazon Revıew Dataset </strong></div>


<div <h1 <strong style="color:#F7C04A;font-size:45px;font-family:Georgia;text-align:center;">RATING REVIEW </strong></div>

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Average </strong></div>
<div <h1 <strong style="color:#F7C04A;font-size:20px;font-family:Georgia;">  Time-Based Weighted Average </strong></div>
<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  User-Based Weighted Average </strong></div>
<div <h1 <strong style="color:#F7C04A;font-size:20px;font-family:Georgia;">  Weighted Rating </strong></div>
<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Bayesian Average Rating Score </strong></div>




In [34]:
import pandas as pd
df = pd.read_csv('/kaggle/input/amazon-reviews/amazon_reviews.csv')


 <div <h1      <strong style="color:#3795BD;font-size:30px;font-family:Georgia;text-align:center;">🤔.DATA UNDERSTANDING.🤔</strong></div>


In [35]:
def check_df(dataframe, head=5):
    print(" SHAPE ".center(70,'-'))
    print('Rows: {}'.format(dataframe.shape[0]))
    print('Columns: {}'.format(dataframe.shape[1]))
    print(" TYPES ".center(70,'-'))
    print(dataframe.dtypes)
    print(" MISSING VALUES ".center(70,'-'))
    print(dataframe.isnull().sum())
    print(" DUPLICATED VALUES ".center(70,'-'))
    print(dataframe.duplicated().sum())
    print(" DESCRIBE ".center(70,'-'))
    print(dataframe.describe())
check_df(df)

------------------------------- SHAPE --------------------------------
Rows: 4915
Columns: 12
------------------------------- TYPES --------------------------------
reviewerID         object
asin               object
reviewerName       object
helpful            object
reviewText         object
overall           float64
summary            object
unixReviewTime      int64
reviewTime         object
day_diff            int64
helpful_yes         int64
total_vote          int64
dtype: object
--------------------------- MISSING VALUES ---------------------------
reviewerID        0
asin              0
reviewerName      1
helpful           0
reviewText        1
overall           0
summary           0
unixReviewTime    0
reviewTime        0
day_diff          0
helpful_yes       0
total_vote        0
dtype: int64
------------------------- DUPLICATED VALUES --------------------------
0
------------------------------ DESCRIBE ------------------------------
         overall   unixReviewTime   day_d

In [36]:
# Let's change the data type of the comment variable to datetime.
df['reviewTime'] = pd.to_datetime(df['reviewTime'])

In [37]:
df["helpful"].head()

0    [0, 0]
1    [0, 0]
2    [0, 0]
3    [0, 0]
4    [0, 0]
Name: helpful, dtype: object

In [38]:
# Let's convert the variable to get the liked and disliked comments.
new_features = df["helpful"].str.split(",", expand=True)
new_features.dtypes

0    object
1    object
dtype: object

In [39]:
new_features.head()

0    1
0  [0   0]
1  [0   0]
2  [0   0]
3  [0   0]
4  [0   0]

In [40]:
#Let's create the variable df["helpful_yes"]
df["helpful_yes"] = new_features[0].str.lstrip("[")
df["helpful_yes"] = df["helpful_yes"].astype("int64")
df["helpful_yes"].head()

0    0
1    0
2    0
3    0
4    0
Name: helpful_yes, dtype: int64

In [41]:
#Let's create the variable df["total_vote"]
df["total_vote"] = new_features[1].str.rstrip("]")
df["total_vote"] = df["total_vote"].astype("int64")
df['total_vote'].head()

0    0
1    0
2    0
3    0
4    0
Name: total_vote, dtype: int64

In [42]:
#Let's create the variable df["helpful_no"]
df["helpful_no"] = df["total_vote"] - df["helpful_yes"]
df["helpful_no"].head()

0    0
1    0
2    0
3    0
4    0
Name: helpful_no, dtype: int64

 <div <h1      <strong style="color:#3795BD;font-size:40px;font-family:Georgia;text-align:center;">Rating Calculation</strong></div>



In [43]:
# Let's observe in order according to Rating.
df.sort_values("overall", ascending=False).head()

reviewerID        asin                         reviewerName helpful                                         reviewText  overall                                            summary  unixReviewTime reviewTime  day_diff  helpful_yes  total_vote  helpful_no
2457  A2HUOPC7F93949  B007WTAJTO  Jonathan R. Adams "adrenalinanimal"  [0, 0]  Was worried about some articles saying to use ...  5.00000  Went into my new Gorpro Hero 3 Black - Camera ...      1374451200 2013-07-22       504            0           0           0
3077   AIO83YNV5QXCU  B007WTAJTO                 M. Birch "Flexstyle"  [0, 0]  I use this in my Samsung Galaxy S3. It's spaci...  5.00000                         Works great with Galaxy S3      1383609600 2013-05-11       576            0           0           0
3083   AR3G5QTJ5GD0T  B007WTAJTO                          McDaddyTree  [0, 0]  Got one for my wife's S3 and it works perfect....  5.00000                                        Works great      1372464000 2013-06-29       527            0           0           0
3082  A14MQ34LD8Q1RT  B007WTAJTO              McClinton Degala "M.D."  [0, 0]  Happy to have gotten this for a steal price! D...  5.00000                             Sales are a good thing      1403740800 2014-06-26       165            0           0           0
3081  A2U55H2MZNIQZM  B007WTAJTO                              MCBazil  [0, 0]  I bought 2 of these, 1 to use in a StealthCam(...  5.00000                                Great microSD cards      1382400000 2013-10-22       412            0           0           0

In [44]:
#The average of the rats will not be enough.
df['overall'].mean()

4.587589013224822

In [45]:
#Look at their weighted averages by point times.
df['reviewTime'] = pd.to_datetime(df['reviewTime'])
current_date = dt.datetime.today()
df['days'] = (current_date - df['reviewTime']).dt.days

In [46]:
# Let's determine the score times, it is possible to set it manually.
a = df['days'].quantile(0.25)
b = df['days'].quantile(0.50)
c = df['days'].quantile(0.75)
# The weights that we will determine in the weighted average may change according to the degree of importance.
def time_based_weighted_average(dataframe, w1=28, w2=26, w3=24, w4=22):
    return dataframe.loc[dataframe["days"] <= a, "overall"].mean() * w1 / 100 + \
           dataframe.loc[(dataframe["days"] > a) & (dataframe["days"] <= b), "overall"].mean() * w2 / 100 + \
           dataframe.loc[(dataframe["days"] > b) & (dataframe["days"] <= c), "overall"].mean() * w3 / 100 + \
           dataframe.loc[(dataframe["days"] > c), "overall"].mean() * w4 / 100

time_based_weighted_average(df)

# We observe that the change we have made in time weights is reflected in the overall score.

4.595593165128118

In [47]:
# Let's execute the wilson_lower_bound function.
def wilson_lower_bound(pos, neg, confidence=0.95):
    import scipy.stats as st
    import math
    n = pos + neg
    if (pos - neg) == 0:
        return 0
    z = st.norm.ppf(1 - (1 - confidence) / 2)
    phat = 1.0 * pos / n
    return (phat + z * z / (2 * n) - z * math.sqrt((phat * (1 - phat) + z * z / (4 * n)) / n)) / (1 + z * z / n)

In [48]:
#And we apply our wilson_lower_bound sorting function.
df["wilson_lower_bound"] = df.apply(lambda x: wilson_lower_bound(x["helpful_yes"], x["helpful_no"]), axis=1)
df.loc[:, ["wilson_lower_bound", 'reviewerID', 'reviewText']].sort_values(by='wilson_lower_bound',
                                                                          ascending=False).head(20)


wilson_lower_bound      reviewerID                                         reviewText
2031             0.95754  A12B7ZMXFI6IXY  [[ UPDATE - 6/19/2014 ]]So my lovely wife boug...
3449             0.93652   AOEAD7DPLZE53  I have tested dozens of SDHC and micro-SDHC ca...
4212             0.91214   AVBMZZAFEKO58  NOTE:  please read the last update (scroll to ...
317              0.81858  A1ZQAQFYSXL5MQ  If your card gets hot enough to be painful, it...
4672             0.80811  A2DKQQIZ793AV5  Sandisk announcement of the first 128GB micro ...
1835             0.78465  A1J6VSUM80UAF8  Bought from BestBuy online the day it was anno...
3981             0.73214  A1K91XXQ6ZEBQR  The last few days I have been diligently shopp...
3807             0.70044   AFGRMORWY2QNX  I bought this card to replace a lost 16 gig in...
4306             0.67033   AOHXKM5URSKAB  While I got this card as a "deal of the day" o...
4596             0.66359  A1WTQUOQ4WG9AI  Hi:I ordered two card and they arrived the nex...
315              0.65741  A2J26NNQX6WKAU  Bought this card to use with my Samsung Galaxy...
1465             0.64567   A6I8KXYK24RTB  I for one have not bought into Google's, or an...
1609             0.64567  A2TPXOZSU1DACQ  I have always been a sandisk guy.  This cards ...
4302             0.63977  A2EL2GWJ9T0DWY  So I got this SD specifically for my GoPro Bla...
4072             0.60967  A22GOZTFA02O2F  I used this for my Samsung Galaxy Tab 2 7.0 . ...
1753             0.56552   ALPLKR59QMBUX  Puchased this card right after I received my S...
1072             0.56552  A2O96COBMVY9C4  What more can I say? The 64GB micro SD works f...
2583             0.56552  A3MEPYZVTAV90W  I bought this Class 10 SD card for my GoPro 3 ...
121              0.56552  A2Z4VVF1NTJWPB  Update: providing an update with regard to San...
1142             0.56552  A1PLHPPAJ5MUXG  As soon as I saw that this card was announced ...

<div <h1 <strong style="color:#F7C04A;font-size:45px;font-family:Georgia;text-align:center;">SORTING REVIEW </strong></div>

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Sorting by Rating </strong></div>
<div <h1 <strong style="color:#F7C04A;font-size:20px;font-family:Georgia;">  Sorting by Comment Count or Purchase Count </strong></div>
<div <h1 <strong style="color:#F7C04A;font-size:20px;font-family:Georgia;">  Sorting by Bayesian Average Rating Score (Sorting Products with 5 Star Rated) </strong></div>
<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Hybrid Sorting: BAR Score + Others </strong></div>



In [49]:
#Sorting by rating
df.sort_values("overall", ascending=True).head()


reviewerID        asin                        reviewerName helpful                                         reviewText  overall                              summary  unixReviewTime reviewTime  day_diff  helpful_yes  total_vote  helpful_no  days  wilson_lower_bound
1365  A1BVU7F2T8EUKS  B007WTAJTO                                DKDC  [2, 4]  Spent over $50 and the thing came fast and did...  1.00000             Maybe I was just unlucky      1347580800 2012-09-14       815            2           4           2  3833             0.00000
1610  A20LC5Z4QSVBZ3  B007WTAJTO            Esteban Esquivel Sanchez  [0, 0]  I hate it, it worked fine for 2 months and the...  1.00000  Worst product I have ever purchased      1364860800 2013-02-04       672            0           0           0  3690             0.00000
3624  A1Y07XHSC38J3T  B007WTAJTO  Phillip W. Wilson "Phillip Wilson"  [1, 2]  I installed this on my Galaxy Note 2 and it cr...  1.00000             Crashed after one month.      1355702400 2012-12-17       721            1           2           1  3739             0.00000
4536  A2POOY3D45I7JR  B007WTAJTO                        Tim Gladding  [1, 2]  Picked this up on sale.. doesn't work, only sh...  1.00000                        Doesn't work.      1352764800 2012-11-13       755            1           2           1  3773             0.00000
3104  A1N2E4RTHC2JPZ  B007WTAJTO                             Melissa  [0, 0]  It worked for a while, I purchased 4 of them, ...  1.00000              It worked for a while..      1401408000 2014-05-30       192            0           0           0  3210             0.00000

In [50]:
#Sorting by Comment
df['comment_count'] = df['helpful_yes'] + df['helpful_yes']
df.sort_values(by="comment_count", ascending=False).head()


reviewerID        asin                reviewerName       helpful                                         reviewText  overall                                            summary  unixReviewTime reviewTime  day_diff  helpful_yes  total_vote  helpful_no  days  wilson_lower_bound  comment_count
2031  A12B7ZMXFI6IXY  B007WTAJTO        Hyoun Kim "Faluzure"  [1952, 2020]  [[ UPDATE - 6/19/2014 ]]So my lovely wife boug...  5.00000  UPDATED - Great w/ Galaxy S4 & Galaxy Tab 4 10...      1367366400 2013-01-05       702         1952        2020          68  3720             0.95754           3904
4212   AVBMZZAFEKO58  B007WTAJTO                 SkincareCEO  [1568, 1694]  NOTE:  please read the last update (scroll to ...  1.00000  1 Star reviews - Micro SDXC card unmounts itse...      1375660800 2013-05-08       579         1568        1694         126  3597             0.91214           3136
3449   AOEAD7DPLZE53  B007WTAJTO           NLee the Engineer  [1428, 1505]  I have tested dozens of SDHC and micro-SDHC ca...  5.00000  Top of the class among all (budget-priced) mic...      1348617600 2012-09-26       803         1428        1505          77  3821             0.93652           2856
317   A1ZQAQFYSXL5MQ  B007WTAJTO     Amazon Customer "Kelly"    [422, 495]  If your card gets hot enough to be painful, it...  1.00000                                Warning, read this!      1346544000 2012-02-09      1033          422         495          73  4051             0.81858            844
3981  A1K91XXQ6ZEBQR  B007WTAJTO  R. Sutton, Jr. "RWSynergy"    [112, 139]  The last few days I have been diligently shopp...  5.00000  Resolving confusion between "Mobile Ultra" and...      1350864000 2012-10-22       777          112         139          27  3795             0.73214            224

In [51]:
# Let's weight it according to the average and the number of comments.
df["comment_count_scaled"] = MinMaxScaler(feature_range=(1, 5)).fit(df[["comment_count"]]).transform(df[["comment_count"]])

def weighted_sorting_score(dataframe, w1=55, w2=45):
    return (dataframe["comment_count_scaled"] * w1 / 100 +
            dataframe["overall"] * w2 / 100)

df["weighted_sorting_score"] = weighted_sorting_score(df)

df.sort_values("weighted_sorting_score", ascending=False).head()



reviewerID        asin                reviewerName       helpful                                         reviewText  overall                                            summary  unixReviewTime reviewTime  day_diff  helpful_yes  total_vote  helpful_no  days  wilson_lower_bound  comment_count  comment_count_scaled  weighted_sorting_score
2031  A12B7ZMXFI6IXY  B007WTAJTO        Hyoun Kim "Faluzure"  [1952, 2020]  [[ UPDATE - 6/19/2014 ]]So my lovely wife boug...  5.00000  UPDATED - Great w/ Galaxy S4 & Galaxy Tab 4 10...      1367366400 2013-01-05       702         1952        2020          68  3720             0.95754           3904               5.00000                 5.00000
3449   AOEAD7DPLZE53  B007WTAJTO           NLee the Engineer  [1428, 1505]  I have tested dozens of SDHC and micro-SDHC ca...  5.00000  Top of the class among all (budget-priced) mic...      1348617600 2012-09-26       803         1428        1505          77  3821             0.93652           2856               3.92623                 4.40943
3981  A1K91XXQ6ZEBQR  B007WTAJTO  R. Sutton, Jr. "RWSynergy"    [112, 139]  The last few days I have been diligently shopp...  5.00000  Resolving confusion between "Mobile Ultra" and...      1350864000 2012-10-22       777          112         139          27  3795             0.73214            224               1.22951                 2.92623
1835  A1J6VSUM80UAF8  B007WTAJTO                 goconfigure      [60, 68]  Bought from BestBuy online the day it was anno...  5.00000                                           I own it      1393545600 2014-02-28       283           60          68           8  3301             0.78465            120               1.12295                 2.86762
4306   AOHXKM5URSKAB  B007WTAJTO               Stellar Eller      [51, 65]  While I got this card as a "deal of the day" o...  5.00000                                      Awesome Card!      1339200000 2012-09-06       823           51          65          14  3841             0.67033            102               1.10451                 2.85748

<div <h1 <strong style="color:#3795BD;font-size:30px;font-family:Georgia;">2 - IMDB Movie Dataset </strong></div>

<div <h1 <strong style="color:#F7C04A;font-size:45px;font-family:Georgia;text-align:center;">RATING PRODUCTS </strong></div>


In [52]:
# Let's upload our new DataFrame for product scores analysis.
movies = pd.read_csv('/kaggle/input/imdb-metadata/movies_metadata.csv', low_memory=False)
movies.head()

adult                              belongs_to_collection    budget                                             genres                              homepage     id    imdb_id original_language               original_title                                           overview popularity                       poster_path                               production_companies                               production_countries release_date         revenue   runtime                                   spoken_languages    status                                            tagline                        title  video  vote_average  vote_count
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  http://toystory.disney.com/toy-story    862  tt0114709                en                    Toy Story  Led by Woody, Andy's toys live happily in his ...  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg     [{'name': 'Pixar Animation Studios', 'id': 3}]  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30 373554033.00000  81.00000           [{'iso_639_1': 'en', 'name': 'English'}]  Released                                                NaN                    Toy Story  False       7.70000  5415.00000
1  False                                                NaN  65000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...                                   NaN   8844  tt0113497                en                      Jumanji  When siblings Judy and Peter discover an encha...  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  [{'name': 'TriStar Pictures', 'id': 559}, {'na...  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15 262797249.00000 104.00000  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released          Roll the dice and unleash the excitement!                      Jumanji  False       6.90000  2413.00000
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...                                   NaN  15602  tt0113228                en             Grumpier Old Men  A family wedding reignites the ancient feud be...    11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22         0.00000 101.00000           [{'iso_639_1': 'en', 'name': 'English'}]  Released  Still Yelling. Still Fighting. Still Ready for...             Grumpier Old Men  False       6.50000    92.00000
3  False                                                NaN  16000000  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...                                   NaN  31357  tt0114885                en            Waiting to Exhale  Cheated on, mistreated and stepped on, the wom...   3.859495  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  [{'name': 'Twentieth Century Fox Film Corporat...  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22  81452156.00000 127.00000           [{'iso_639_1': 'en', 'name': 'English'}]  Released  Friends are the people who let you be yourself...            Waiting to Exhale  False       6.10000    34.00000
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0                     [{'id': 35, 'name': 'Comedy'}]                                   NaN  11862  tt0113041                en  Father of the Bride Part II  Just when George Banks has recovered from his ...   8.387519  /e64sOI48hQXyru7naBFyssKFxVd.jpg  [{'name': 'Sandollar Productions', 'id': 5842}...  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10  76578911.00000 106.00000           [{'iso_639_1': 'en', 'name': 'English'}]  Released  Just When His World Is Back To Normal... He's ...  Father of the Bride Part II  False       5.70000   173.00000

 <div <h1      <strong style="color:#3795BD;font-size:30px;font-family:Georgia;text-align:center;">🤔.DATA UNDERSTANDING.🤔</strong></div>


In [53]:
movies = movies[["title", "vote_average", "vote_count"]]

def check_df(dataframe, head=5):
    print(" SHAPE ".center(70,'-'))
    print('Rows: {}'.format(dataframe.shape[0]))
    print('Columns: {}'.format(dataframe.shape[1]))
    print(" TYPES ".center(70,'-'))
    print(dataframe.dtypes)
    print(" HEAD ".center(70,'-'))
    print(dataframe.head(head))
    print(" MISSING VALUES ".center(70,'-'))
    print(dataframe.isnull().sum())
    print(" DUPLICATED VALUES ".center(70,'-'))
    print(dataframe.duplicated().sum())
    print(" DESCRIBE ".center(70,'-'))
    print(dataframe.describe().T)
check_df(movies)


------------------------------- SHAPE --------------------------------
Rows: 45466
Columns: 3
------------------------------- TYPES --------------------------------
title            object
vote_average    float64
vote_count      float64
dtype: object
-------------------------------- HEAD --------------------------------
                         title  vote_average  vote_count
0                    Toy Story       7.70000  5415.00000
1                      Jumanji       6.90000  2413.00000
2             Grumpier Old Men       6.50000    92.00000
3            Waiting to Exhale       6.10000    34.00000
4  Father of the Bride Part II       5.70000   173.00000
--------------------------- MISSING VALUES ---------------------------
title           6
vote_average    6
vote_count      6
dtype: int64
------------------------- DUPLICATED VALUES --------------------------
52
------------------------------ DESCRIBE ------------------------------
                   count      mean       std     min 

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Sorting by Rating </strong></div>

In [54]:
# Watch the movies that come in order according to the movie score_average.
movies.sort_values("vote_average", ascending=False).head(20)


title  vote_average  vote_count
21642  Ice Age Columbus: Who Were the First Americans?      10.00000     1.00000
15710        If God Is Willing and da Creek Don't Rise      10.00000     1.00000
22396                                   Meat the Truth      10.00000     1.00000
22395            Marvin Hamlisch: What He Did For Love      10.00000     1.00000
35343                       Elaine Stritch: At Liberty      10.00000     1.00000
186                                           Reckless      10.00000     1.00000
45047                                  The Human Surge      10.00000     1.00000
22377                                        The Guide      10.00000     1.00000
22346                                   هیچ کجا هیچ کس      10.00000     1.00000
1634                          Other Voices Other Rooms      10.00000     1.00000
35461                          Tom Brown's School Days      10.00000     1.00000
22255                                  Cowboy del Amor      10.00000     1.00000
45005                                       Fool's Day      10.00000     1.00000
41879                   Panorama of Esplanade by Night      10.00000     1.00000
41845                                        Maula Jat      10.00000     1.00000
40431                              People of a Feather      10.00000     1.00000
40442                              Siccin 3: Cürmü Ask      10.00000     2.00000
27908                                           Mentor      10.00000     1.00000
41804                                       Bella Vita      10.00000     1.00000
41797      Campaign of Hate: Russia and Gay Propaganda      10.00000     1.00000

In [55]:
movies["vote_count"].describe([0.10, 0.25, 0.50, 0.70, 0.80, 0.90, 0.95, 0.99]).T

count   45460.00000
mean      109.89734
std       491.31037
min         0.00000
10%         1.00000
25%         3.00000
50%        10.00000
70%        25.00000
80%        50.00000
90%       160.00000
95%       434.00000
99%      2183.82000
max     14075.00000
Name: vote_count, dtype: float64

In [56]:
## Oylama sayisina gore filtre koyup tekrar gozlemleyelim
movies[movies["vote_count"] > 400].sort_values("vote_average", ascending=False).head(20)

title  vote_average  vote_count
10309      Dilwale Dulhania Le Jayenge       9.10000   661.00000
40251                       Your Name.       8.50000  1030.00000
834                      The Godfather       8.50000  6024.00000
314           The Shawshank Redemption       8.50000  8358.00000
1152   One Flew Over the Cuckoo's Nest       8.30000  3001.00000
1176                            Psycho       8.30000  2405.00000
1178            The Godfather: Part II       8.30000  3418.00000
292                       Pulp Fiction       8.30000  8670.00000
1184       Once Upon a Time in America       8.30000  1104.00000
5481                     Spirited Away       8.30000  3968.00000
522                   Schindler's List       8.30000  4436.00000
23673                         Whiplash       8.30000  4376.00000
2843                        Fight Club       8.30000  9678.00000
2211                 Life Is Beautiful       8.30000  3643.00000
12481                  The Dark Knight       8.30000 12269.00000
5878                       City of God       8.20000  1852.00000
1170                        GoodFellas       8.20000  3211.00000
1161                      12 Angry Men       8.20000  2130.00000
351                       Forrest Gump       8.20000  8147.00000
39386                            Piper       8.20000   487.00000

In [57]:
#Let's update the vote_count variable to values between 1-10 and list it again.
movies["vote_count_score"] = MinMaxScaler(feature_range=(1, 10)).fit(movies[["vote_count"]]).transform(movies[["vote_count"]])

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  vote_average * vote_count </strong></div>

In [58]:
# Let's create our score variable by multiplying our vote_average and vote_count_score variables in the same format.
movies["average_count_score"] = movies["vote_average"] * movies["vote_count_score"]

movies.sort_values("average_count_score", ascending=False).head(20).style.set_properties(**{"background-color": "#000000","color":"#FFEA20","border": "0.2px solid white"})

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  weighted_rating = (v/(v+M) * r) + (M/(v+M) * C) </strong></div>

* r = vote average
* v = vote count
* M = minimum votes required to be listed in the Top 250
* C = the mean vote across the whole report (currently 7.0)

In [59]:
#Let's apply our formula to our variable with a score of 8.5 and a vote count of 8358 as follows.
def weighted_rating(r, v, M, C):
    return (v / (v + M) * r) + (M / (v + M) * C)

M = 2500
C=5.6

weighted_rating(8.50, 8358, M, C)

7.832289556087677

In [60]:
movies.sort_values("average_count_score", ascending=False).head(10).style.set_properties(**{"background-color": "#000000","color":"#FFEA20","border": "0.2px solid white"})


In [61]:
# Observe by applying our weighted scores to our entire table.

movies["weighted_rating"] = weighted_rating(movies["vote_average"],
                                        movies["vote_count"], M, C)

movies.sort_values("weighted_rating", ascending=False).head(10).style.set_properties(**{"background-color": "#000000","color":"#FFEA20","border": "0.2px solid white"})

<div <h1 <strong style="color:#3795BD;font-size:20px;font-family:Georgia;">  Bayesian Average Rating Score </strong></div>

*  12481..........                                    The Dark Knight
*  314 ..........                            The Shawshank Redemption
*  2843 .........                                         Fight Club
*  15480  ......                                        Inception
*  292  .........                                       Pulp Fiction

In [62]:
# confidence=0.95  --> % 95
def bayesian_average_rating(n, confidence=0.95):
    if sum(n) == 0:
        return 0
    K = len(n)
    z = st.norm.ppf(1 - (1 - confidence) / 2)
    N = sum(n)
    first_part = 0.0
    second_part = 0.0
    for k, n_k in enumerate(n):
        first_part += (k + 1) * (n[k] + 1) / (N + K)
        second_part += (k + 1) * (k + 1) * (n[k] + 1) / (N + K)
    score = first_part - z * math.sqrt((second_part - first_part * first_part) / (N + K + 1))
    return score


# bayesian_average_rating formulunun asagidaki aldigi degerler yildiz sayilarini ifade etmektedir.
# bayesian_average_rating(1_yildiz_sayisi, 2_yildiz_sayisi......)
# Asagidaki islemlerde daha iyi gozlemlenebilecektir

bayesian_average_rating([34733, 4355, 4704, 6561, 13515, 26183, 87368, 273082, 600260, 1295351])

9.14538444560111

In [63]:
movies2 = pd.read_csv("/kaggle/input/imdb-ratings/imdb_ratings.csv")
movies2 = movies2.iloc[0:, 1:]

In [64]:
movies2["bar_score"] = movies2.apply(lambda x: bayesian_average_rating(x[["one", "two", "three", "four", "five",
                                                                "six", "seven", "eight", "nine", "ten"]]), axis=1)
movies2.sort_values("bar_score", ascending=False).head(20).style.set_properties(**{"background-color": "#000000","color":"#FFEA20","border": "0.2px solid white"})

<p style="padding:15px;
background-color:#f4ebdc;
margin:0; color:#b26960; border:2px dotted #C689C6; font-family:Charter; font-weight: bold; font-size:250%; text-align:center; overflow:hidden; font-weight:500">If you like this then please UPVOTE this 😄 Your opinions and suggestions are very important to me...<img src="https://media.giphy.com/media/WUlplcMpOCEmTGBtBW/giphy.gif" width="100"> <div style="float:center;margin-left:270px;max-width:50%;">


<p style="padding:15px;
background-color:#f4ebdc;
margin:0; color:#b26960; border:2px dotted #C689C6; font-family:Charter; font-weight: bold; font-size:250%; text-align:center; overflow:hidden; font-weight:500">
FOR MORE:</p> 

https://www.linkedin.com/in/serdar-ozturk/

https://github.com/StanleyHopson

https://medium.com/@serdar.f95

<p style="padding:15px;
background-color:#f4ebdc;
margin:0; color:#b26960; border:2px dotted #C689C6; font-family:Charter; font-weight: bold; font-size:250%; text-align:center; overflow:hidden; font-weight:500">CREDITS:</p> 

[https://numpy.org/](http://)

[https://seaborn.pydata.org/](http://)

[https://pandas.pydata.org](http://)

[https://pandas.pydata.org/](http://)

[https://learning.miuul.com](http://)